File: app/src/main/java/ocs/com/dr_tips/fragment/PickImageDialogFragment.java
package ocs.com.dr_tips.fragment;

/**
 * Created by sherif on 05/12/17.
 */

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import ocs.com.dr_tips.R;

/**
 * Used to display dialog for user to select Camera or Gallery for selecting images
 */
public class PickImageDialogFragment extends DialogFragment {
    Unbinder unbinder;
    private OnActionPickedListener mListener;

    public PickImageDialogFragment() {
        // Required empty public constructor
    }

    public static PickImageDialogFragment newInstance(OnActionPickedListener listener) {
        PickImageDialogFragment fragment = new PickImageDialogFragment();
        fragment.setOnActionPickedListenerListener(listener);
        return fragment;
    }

    @OnClick(R.id.Btn_Dialog_Image_Picker_Camera)
    public void openCamera() {
        if (mListener != null) {
            mListener.onActionPicked(true);
        }
        dismiss();
    }

    @OnClick(R.id.Btn_Dialog_Image_Picker_Gallery)
    public void openGallery() {
        if (mListener != null) {
            mListener.onActionPicked(false);
        }
        dismiss();
    }

    public void setOnActionPickedListenerListener(OnActionPickedListener mListener) {
        this.mListener = mListener;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_image_picker_dialog, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    public interface OnActionPickedListener {
        void onActionPicked(boolean isCameraChosen);
    }
}

File: app/src/main/java/ocs/com/dr_tips/fragment/PickImageDialogFragment.kt
package ocs.com.dr_tips.fragment

/**
 * Created by sherif on 05/12/17.
 */

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.provider.MediaStore
import android.support.v13.app.ActivityCompat
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.assist.FailReason
import com.nostra13.universalimageloader.core.listener.ImageLoadingListener
import kotlinx.android.synthetic.main.fragment_image_picker_dialog.*
import ocs.com.dr_tips.DrTipsApplication
import ocs.com.dr_tips.R
import ocs.com.dr_tips.util.Utils
import javax.inject.Inject

/**
 * Used to display dialog for user to select Camera or Gallery for selecting images
 */
class PickImageDialogFragment : DialogFragment() {
    private var mListener: OnImagePickedListener? = null
    private val CAMERA_REQUEST: Int = 1
    private val GALLERY_REQUEST: Int = 2
    @Inject
    lateinit var imageLoader: ImageLoader

    fun setOnActionPickedListenerListener(mListener: OnImagePickedListener) {
        this.mListener = mListener
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as DrTipsApplication).component.inject(this)
    }
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? = inflater.inflate(R.layout.fragment_image_picker_dialog, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Btn_Dialog_Image_Picker_Camera.setOnClickListener {
            if (askForWriteExternalStoragePermission()) {
                openCamera()
            }
        }
        Btn_Dialog_Image_Picker_Gallery.setOnClickListener {
            if (askForWriteExternalStoragePermission()) {
                openGallery()
            }
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    interface OnImagePickedListener {
        fun onImagePicked(selectedImage: Bitmap)
    }

    private fun openGallery() {
        val intent = Intent()
        intent.type = "image/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(Intent.createChooser(intent, "Select Photo"), GALLERY_REQUEST)
    }

    private fun openCamera() {
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        startActivityForResult(intent, CAMERA_REQUEST)
    }

    companion object {
        fun newInstance(listener: OnImagePickedListener): PickImageDialogFragment {
            val fragment = PickImageDialogFragment()
            fragment.setOnActionPickedListenerListener(listener)
            return fragment
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == CAMERA_REQUEST) {
                Utils.fixMediaDir()
                val profilePic = data?.extras?.get("data") as Bitmap
                mListener?.onImagePicked(profilePic)
                dismiss()
            } else if (requestCode == GALLERY_REQUEST) {
                val imageUri = data?.data
                imageLoader.loadImage(imageUri.toString(), object : ImageLoadingListener {
                    override fun onLoadingComplete(imageUri: String?, view: View?, loadedImage: Bitmap?) {
                        mListener?.onImagePicked(loadedImage!!)
                        dismiss()
                    }

                    override fun onLoadingStarted(imageUri: String?, view: View?) {
                    }

                    override fun onLoadingCancelled(imageUri: String?, view: View?) {
                    }

                    override fun onLoadingFailed(imageUri: String?, view: View?, failReason: FailReason?) {
                    }

                })
            }
        }
    }

    fun askForWriteExternalStoragePermission(): Boolean {
        if ((ActivityCompat.shouldShowRequestPermissionRationale(activity!!, Manifest.permission.WRITE_EXTERNAL_STORAGE))||
                (ActivityCompat.shouldShowRequestPermissionRationale(activity!!, Manifest.permission.CAMERA)) ) {
            ActivityCompat.requestPermissions(activity!!,
                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.CAMERA), 1)
            return false
        }
        return true
    }
}

